File: Laborator/CSharp/AgentieTurism/Repository/FlightDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AgentieTurism.Models;
using AgentieTurism.Utils;
using log4net;

namespace AgentieTurism.Repository
{
    public class FlightDbRepository : IFlightRepository
    {
        private static readonly ILog log = LogManager.GetLogger("FlightDbRepo");
        //private readonly IDictionary<string, string> _properties;

        //public FlightDbRepository(IDictionary<string, string> props)
        public FlightDbRepository()
        {
            log.Info("Initializing FlightDbRepository");
            //_properties = props;
        }

        public void Add(Flight flight)
        {
            log.Info($"Adding flight: {flight}");
            var sql = "INSERT INTO Flights (destination, departureDateTime, airport, availableSeats) VALUES (@Destination, @DepartureDateTime, @Airport, @AvailableSeats)";
            var parameters = new Dictionary<string, object>
            {
                { "@Destination", flight.Destination },
                { "@DepartureDateTime", flight.DepartureDateTime.ToString("yyyy-MM-dd HH:mm") },
                { "@Airport", flight.Airport },
                { "@AvailableSeats", flight.AvailableSeats }
            };


            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    foreach (var param in parameters)
                    {
                        var dbParam = cmd.CreateParameter();
                        dbParam.ParameterName = param.Key;
                        dbParam.Value = param.Value;
                        cmd.Parameters.Add(dbParam);
                    }

                    cmd.ExecuteNonQuery();

                    cmd.CommandText = "SELECT last_insert_rowid()";
                    flight.Id = Convert.ToInt32(cmd.ExecuteScalar());
                }
                log.Info($"Flight to {flight.Destination} added successfully");
            }
            catch (Exception ex)
            {
                log.Error("Failed to add flight", ex);
            }
        }

        public void Update(int id, Flight flight)
        {
            log.Info($"Updating flight with ID: {id}");
            var sql = "UPDATE Flights SET destination = @Destination, departureDateTime = @DepartureDateTime, airport = @Airport, availableSeats = @AvailableSeats WHERE id = @Id";
            var parameters = new Dictionary<string, object>
            {
                { "@Id", id },
                { "@Destination", flight.Destination },
                { "@DepartureDateTime", flight.DepartureDateTime.ToString("yyyy-MM-dd HH:mm") },
                { "@Airport", flight.Airport },
                { "@AvailableSeats", flight.AvailableSeats }
            };
            ExecuteNonQuery(sql, parameters);
            log.Info($"Flight with ID: {id} updated successfully");
        }

        public void Delete(int id)
        {
            log.Info($"Deleting flight with ID: {id}");
            var sql = "DELETE FROM Flights WHERE id = @Id";
            var parameters = new Dictionary<string, object> { { "@Id", id } };
            ExecuteNonQuery(sql, parameters);
            log.Info($"Flight with ID: {id} deleted successfully");
        }

        public Flight FindOne(int id)
        {
            log.Info($"Finding flight by ID: {id}");
            var sql = "SELECT * FROM Flights WHERE id = @Id";
            var parameters = new Dictionary<string, object> { { "@Id", id } };
            var flight = ExecuteQuerySingle(sql, parameters, DecodeFlight);
            log.Info(flight != null ? $"Flight found: {flight}" : "Flight not found");
            return flight;
        }

        public List<Flight> FindAll()
        {
            log.Info("Retrieving all flights");
            var sql = "SELECT * FROM Flights";
            var flights = ExecuteQueryList(sql, null, DecodeFlight);
            log.Info($"Found {flights.Count} flights");
            return flights;
        }

        public List<Flight> FindFlights()
        {
            log.Info("Retrieving all flights");
            var sql = "SELECT * FROM Flights WHERE availableSeats > 0";
            var flights = ExecuteQueryList(sql, null, DecodeFlight);
            log.Info($"Found {flights.Count} flights");
            return flights;
        }

        public List<Flight> FindByDestinationAndDateTime(string destination, DateTime departureDate)
        {
            log.Info($"Finding flights by destination: '{destination}' and date: {departureDate:yyyy-MM-dd}");

            var sql = @"
                SELECT * FROM Flights
                WHERE destination LIKE @Destination
                AND SUBSTR(departureDateTime, 1, 10) = @DepartureDate";

            var parameters = new Dictionary<string, object>
        {
            { "@Destination", $"%{destination}%" },
            { "@DepartureDate", departureDate.ToString("yyyy-MM-dd") }
        };

            return ExecuteQueryList(sql, parameters, DecodeFlight);
        }

        public void UpdateAvailableSeats(int flightId, int newAvailableSeats)
        {
            log.Info($"Updating available seats for flight ID: {flightId}");
            var sql = "UPDATE Flights SET availableSeats = @AvailableSeats WHERE id = @FlightId";
            var parameters = new Dictionary<string, object>
            {
                { "@FlightId", flightId },
                { "@AvailableSeats", newAvailableSeats }
            };
            ExecuteNonQuery(sql, parameters);
            log.Info($"Updated available seats for flight ID: {flightId} to {newAvailableSeats}");
        }

        private Flight DecodeFlight(IDataReader reader)
        {
            Flight flight = new Flight(
                reader.GetString(1),
                reader.GetDateTime(2),
                reader.GetString(3),
                reader.GetInt32(4)
            );
            flight.Id = reader.GetInt32(0);

            return flight;
        }

        private void ExecuteNonQuery(string sql, Dictionary<string, object> parameters)
        {
            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    foreach (var param in parameters)
                    {
                        var dbParam = cmd.CreateParameter();
                        dbParam.ParameterName = param.Key;
                        dbParam.Value = param.Value;
                        cmd.Parameters.Add(dbParam);
                    }
                    cmd.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                log.Error("Database operation failed", ex);
            }
        }

        private Flight ExecuteQuerySingle(string sql, Dictionary<string, object> parameters, Func<IDataReader, Flight> decoder)
        {
            try
            {
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    if (parameters != null)
                    {
                        foreach (var param in parameters)
                        {
                            var dbParam = cmd.CreateParameter();
                            dbParam.ParameterName = param.Key;
                            dbParam.Value = param.Value;
                            cmd.Parameters.Add(dbParam);
                        }
                    }

                    using (var reader = cmd.ExecuteReader())
                    {
                        return reader.Read() ? decoder(reader) : null;
                    }
                }
            }
            catch (Exception ex)
            {
                log.Error("Database query failed", ex);
                return null;
            }
        }

        private List<Flight> ExecuteQueryList(string sql, Dictionary<string, object> parameters, Func<IDataReader, Flight> decoder)
        {
            try
            {
                var flights = new List<Flight>();
                var con = DbConnectionUtils.GetConnection();
                using (var cmd = con.CreateCommand())
                {
                    cmd.CommandText = sql;
                    if (parameters != null)
                    {
                        foreach (var param in parameters)
                        {
                            var dbParam = cmd.CreateParameter();
                            dbParam.ParameterName = param.Key;
                            dbParam.Value = param.Value;
                            cmd.Parameters.Add(dbParam);
                        }
                    }

                    using (var reader = cmd.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            flights.Add(decoder(reader));
                        }
                    }
                }
                return flights;
            }
            catch (Exception ex)
            {
                log.Error("Database query failed", ex);
                return new List<Flight>();
            }
        }
    }
}

File: Laborator/Java/AgentieTurism/src/main/java/ro/mpp2024/repository/FlightDBRepository.java
package ro.mpp2024.repository;

import ro.mpp2024.utils.JdbcUtils;
import ro.mpp2024.domain.Flight;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class FlightDBRepository implements FlightRepository {
    private JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public FlightDBRepository(Properties props) {
        logger.info("Initializing FlightDBRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void add(Flight flight) {
        logger.traceEntry("Saving flight {}", flight);
        String sql = "INSERT INTO Flights (destination, departureDateTime, airport, availableSeats) VALUES (?, ?, ?, ?)";

        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
            stmt.setString(1, flight.getDestination());
            stmt.setString(2, flight.getDepartureDateTime().format(formatter));
            stmt.setString(3, flight.getAirport());
            stmt.setInt(4, flight.getAvailableSeats());

            int affectedRows = stmt.executeUpdate();
            if (affectedRows == 0) {
                throw new SQLException("Creating flight failed, no rows affected.");
            }

            try (ResultSet generatedKeys = stmt.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    int generatedId = generatedKeys.getInt(1);
                    flight.setId(generatedId);
                    logger.trace("Flight saved successfully with ID: {}", generatedId);
                } else {
                    throw new SQLException("Creating flight failed, no ID obtained.");
                }
            }
        } catch (SQLException e) {
            logger.error("Error inserting flight", e);
        }
        logger.traceExit();
    }


    @Override
    public void update(Integer id, Flight flight) {
        logger.traceEntry("Updating flight with ID: {} to {}", id, flight);
        String sql = "UPDATE Flights SET destination = ?, departureDateTime = ?, airport = ?, availableSeats = ? WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setString(1, flight.getDestination());
            stmt.setString(2, flight.getDepartureDateTime().format(formatter));
            stmt.setString(3, flight.getAirport());
            stmt.setInt(4, flight.getAvailableSeats());
            stmt.setInt(5, id);
            stmt.executeUpdate();
            logger.trace("Flight updated successfully.");
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer id) {
        logger.traceEntry("Deleting flight with ID: {}", id);
        String sql = "DELETE FROM Flights WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setInt(1, id);
            int result = stmt.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }


    @Override
    public Flight findOne(Integer id) {
        logger.traceEntry("Finding flight by ID: {}", id);
        String sql = "SELECT * FROM Flights WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setInt(1, id);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Flight flight = extractFlightFromResultSet(rs);
                logger.traceExit(flight);
                return flight;
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        return null;
    }

    @Override
    public Iterable<Flight> findAll() {
        logger.traceEntry();
        List<Flight> flights = new ArrayList<>();
        String sql = "SELECT * FROM Flights";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql);
             ResultSet rs = stmt.executeQuery()) {
            while (rs.next()) {
                flights.add(extractFlightFromResultSet(rs));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit(flights);
        return flights;
    }

    public List<Flight> findFlights() {
        logger.traceEntry();
        List<Flight> flights = new ArrayList<>();
        String sql = "SELECT * FROM Flights WHERE availableSeats > 0";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql);
             ResultSet rs = stmt.executeQuery()) {
            while (rs.next()) {
                flights.add(extractFlightFromResultSet(rs));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit(flights);
        return flights;
    }

//    @Override
//    public List<Flight> findByDestinationAndDateTime(String destination, LocalDateTime departureDateTime) {
//        logger.traceEntry("Finding flights by destination: {} and dateTime: {}", destination, departureDateTime);
//        List<Flight> flights = new ArrayList<>();
//        String sql = "SELECT * FROM Flights WHERE destination = ? AND departureDateTime = ?";
//        try (Connection con = dbUtils.getConnection();
//             PreparedStatement stmt = con.prepareStatement(sql)) {
//            stmt.setString(1, destination);
//            stmt.setString(2, departureDateTime.format(formatter));
//            ResultSet rs = stmt.executeQuery();
//            while (rs.next()) {
//                flights.add(extractFlightFromResultSet(rs));
//            }
//        } catch (SQLException e) {
//            logger.error(e);
//        }
//        logger.traceExit(flights);
//        return flights;
//    }

    @Override
    public List<Flight> findByDestinationAndDateTime(String destination, LocalDateTime departureDateTime) {
        logger.traceEntry("Finding flights by destination: {} and date: {}", destination, departureDateTime.toLocalDate());
        List<Flight> flights = new ArrayList<>();

        String sql = "SELECT * FROM Flights WHERE destination = ? AND SUBSTR(departureDateTime, 1, 10) = ?";

        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setString(1, destination);
            stmt.setString(2, departureDateTime.toLocalDate().toString());

            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                flights.add(extractFlightFromResultSet(rs));
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit(flights);
        return flights;
    }

    @Override
    public void updateAvailableSeats(Integer flightId, int newAvailableSeats) {
        logger.traceEntry("Updating available seats for flight ID: {}", flightId);
        String sql = "UPDATE Flights SET availableSeats = ? WHERE id = ?";
        try (Connection con = dbUtils.getConnection();
             PreparedStatement stmt = con.prepareStatement(sql)) {
            stmt.setInt(1, newAvailableSeats);
            stmt.setInt(2, flightId);
            stmt.executeUpdate();
            logger.trace("Updated available seats successfully.");
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    private Flight extractFlightFromResultSet(ResultSet rs) throws SQLException {

        LocalDateTime departureDateTime = LocalDateTime.parse(rs.getString("departureDateTime"), formatter);
        Flight flight = new Flight(
                rs.getString("destination"),
                departureDateTime,
                rs.getString("airport"),
                rs.getInt("availableSeats")
        );
        flight.setId(rs.getInt("id"));
        return flight;
    }
}
